weakmap.c

```diff
@@ -398,7 +432,30 @@ wmap_aset_replace(st_data_t *key, st_data_t *val, st_data_t new_key_ptr, int exi
     return ST_CONTINUE;
 }
 
-/* Creates a weak reference from the given key to the given value */
+/*
+ *  call-seq:
+ *    map[key] = value -> value
+ *
+ *  Associates the given +value+ with the given +key+; returns +value+.
+ *
+ *  The reference to both +key+ and +value+ are weak, so when there is no
+ *  other references to +key+ or +value+ left, they may be garbage collected.
+ *
+ *     m = ObjectSpace::WeekMap.new
+ *     key = "foo"
+ *     val = Object.new
+ *     m[key] = val
+ *
+ *     m[key] #=> #<Object:0x00007fea7156c580>
+ *
+ *     val = nil # remove the other reference to value
+ *     GC.start
+ *
+ *     m[key] #=> nil
```

Comment:
Note that since Ruby GC is conservative, this is not at all guaranteed to work as showcased. So I'd rather not include this kind of examples with `GC.start`.

Reply:
I'd prefer something that say objects are never guaranteed to be GCed.

Reply:
@byroot What's your preferred wording then? 

I still feel it would be good to leave some examples. 
I am open to any suggestions that would make them less misleading, though I think that `GC.start` (which would work in a simple test script to "experiment how it works") + an explanation that `GC.start` is only for demo purposes might be enough.

Reply:
IMO a comment that say there is no longer any strong reference on that object is enough. I really don't like examples with `GC.start` it tend to really confuse people.

Reply:
That said, I don't consider myself good at documentation, so do what you want, I don't have a very strong opinion here.

Reply:
@byroot It is just from all my experience of learning and teaching programming, the actual code examples (however artificial they are) for many people are telling much more than the most careful wording. 

And I think everybody understands it is artificial (and a note that "it is just for demonstrational purposes" would further clarify that), just like most people understand that examples showing `hash["foo"] = "bar"` don't claim "you should work with `"foo"` and `"bar"` in your real code."

`GC.start` is the only way to demonstrate the _actual behavior_ of the "weak maps" (and how they are different from "just maps") and can be invaluable for the initial understanding.

We had a similar discussion when designing [Ractor](https://docs.ruby-lang.org/en/master/Ractor.html) docs and eventually agreed [on introducing](https://docs.ruby-lang.org/en/master/Ractor.html#class-Ractor-label-Note+on+code+examples) very artificial (and, in general, non-deterministic) `wait` method, which, though, allowed to write artificial yet very demonstrative code.

So... Is there a chance you'll approve it in this form?.. (Considering I've already added the notice about `GC.start` to every example using it.)

Reply:
> So... Is there a chance you'll approve it in this form?.

You don't need my approval. You asked for my opinion, I gave it, but I won't approve nor block documentation changes.